roleccl: enable GRANT/REVOKE for roles without a license

knz opened the pull request:
Fixes #45275.
This was discussed (and agreed upon) with @nstewart and @piyush-singh.

Release note (security update): Non-licensed users are now
able to add more principals to the special superuser role/group
`admin`. (Creation of additional roles is still a licensed feature.)

Release note (sql change): It is now possible to use `GRANT` and
`REVOKE` to add users to the `admin` role without a valid
license. This change aims to enable use of the admin UI and other
privileged features without a license.

knz committed:
roleccl: enable GRANT/REVOKE for roles without a license

Release note (security update): Non-licensed users are now
able to add more principals to the special superuser role/group
`admin`. (Creation of additional roles is still a licensed feature).

Release note (sql change): It is now possible to use `GRANT` and
`REVOKE` to add users to the `admin` role without a valid
license. This change aims to enable use of the admin UI and other
privileged features without a license.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45325)
<!-- Reviewable:end -->

knz commented:
cc @mattcrdb

knz commented:
cc @taroface  - this change will also be backported so the corresponding doc update will be multi-version.

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @RichardJCai, and @rohany)

___
*[pkg/ccl/roleccl/role.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45325#-M0s5PNM7JtC9EPw78gs:-M0s5PNM7JtC9EPw78gt:b-du2yzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d6eb385f681c3d24fc0d6dca1ca106b5817aba/pkg/ccl/roleccl/role.go#L85)):*
> ```Go
> 		// non-licensed users to add/remove users from the admin role, so
> 		// they can grant administrative privileges to user accounts that
> 		// are not superusers like "root".
> ```

Nit: Are we concerned at all about users going from enterprise to non-enterprise? They would be able to grant existing roles in that case.


<!-- Sent from Reviewable.io -->

knz commented:
> Nit: Are we concerned at all about users going from enterprise to non-enterprise? They would be able to grant existing roles in that case.

Good question. Asking @nstewart:

> No enterprise customers are paying JUST for RBAC, this is not a risk IMO. Furthermore, if they were big enough to pay our license for roles, they would certainly want to be able to edit them

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @rohany)

___
*[pkg/ccl/roleccl/role.go, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45325#-M0s7dcK6sqXFJTQI6N2:-M0s7dcK6sqXFJTQI6N3:b-rajoi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d6eb385f681c3d24fc0d6dca1ca106b5817aba/pkg/ccl/roleccl/role.go#L236)):*
> ```Go
> 
> 		if err := utilccl.CheckEnterpriseEnabled(
> 			p.ExecCfg().Settings, p.ExecCfg().ClusterID(), p.ExecCfg().Organization(), "REVOKE <role>",
> ```

Need to remove this check too.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @rohany)

___
*[pkg/ccl/roleccl/role.go, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45325#-M0s7dcK6sqXFJTQI6N2:-M0s8HL-9msdE374jQE5:b-k674xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d6eb385f681c3d24fc0d6dca1ca106b5817aba/pkg/ccl/roleccl/role.go#L236)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Need to remove this check too.
</blockquote></details>

Also probably want to add tests to sql/logictest/testdata/logic_test/role


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45325)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @RichardJCai, and @rohany)

___
*[pkg/ccl/roleccl/role.go, line 236 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45325#-M0s7dcK6sqXFJTQI6N2:-M0s8Xxh8AKnIBMF-hmy:b-l30q3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/35d6eb385f681c3d24fc0d6dca1ca106b5817aba/pkg/ccl/roleccl/role.go#L236)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Also probably want to add tests to sql/logictest/testdata/logic_test/role
</blockquote></details>

Good catch!

The GRANT and REVOKE statements are already tested in that logic test file.

Unfortunately we do not have testing infrastructure for checking that licensed features are available to non-licensed users and vice-versa. As this PR will be backported, I am reluctant to add new infrastructure as it would make the backport more complicated to manage.  (This would be an exception of course.)

I'll file an issue.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45325#-:-M0sPU_F4qA_EDsgPo8U:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @rohany)



<!-- Sent from Reviewable.io -->

taroface commented:
> cc @taroface - this change will also be backported so the corresponding doc update will be multi-version.

@knz Which CRDB versions will receive this update?

knz commented:
> Which CRDB versions will receive this update?

19.1 and 19.2 since the previous changes introduced a serious UX regression.
Probably not 2.1 for now but if there's customer pressure we could consider.

knz committed:
server: return HTTP 403 instead of 500 upon authz failures

Before, the HTTP endpoints would return 500 (Internal server error)
when they require an admin user and a non-admin user was logged in.
This patch changes it to make them return 403 (Forbidden) instead,
which is the standard "permission denied" error code.

Release note (general change): HTTP endpoints now report status
403 (Forbidden) instead of 500 (Internal server error) when the
authenticated user has insufficient privileges to use the endpoint.
